puzzle-utils/words/ladder.py

```diff
@@ -0,0 +1,166 @@
+        return []
+
+    # Find words closer to destination (avoiding cycles)
+    max_edit_distance = edit_distance(current_word, right_word) if use_heuristic else len(current_word)
+    adjacent_words = set()
+    if current_word in cached_searches.keys():
+        adjacent_words = cached_searches[current_word]
+    else:
+        adjacent_words = ladder(current_word)
+        cached_searches[current_word] = adjacent_words
+    used_words = set(word_path)
+    next_words = list(filter(lambda w: w not in used_words and edit_distance(w, right_word) <= max_edit_distance, adjacent_words))
+
+    # No ladder exists this way
+    if len(next_words) == 0:
+        return []
+
+    found_paths = []
+    for next_word in next_words:
+        word_path.append(next_word)
+        results = follow_ladder_path(word_path, right_word, cached_searches, depth + 1, exact_depth, max_depth, use_heuristic)
+        word_path.pop()
+
+        # We have one or more paths to examine
+        if len(results) != 0:
+            for path in results:
+                found_paths.append(path)
+
+    return found_paths
+
+def find_ladders(left_word: str, right_word: str, exact_depth: Optional[int] = None, max_depth: int = 10, use_heuristic: bool = True) -> List[List[str]]:
+    """ Find a word ladder
+
+    Given two words, find every "path" from left_word to right_word
+    Each path is a list of words where each word is an edit distance
+    of 1 from each neighbour. The list starts with left_word and ends
+    with right_word
+
+    Args:
+        left_word: The start word
+        right_word: The end word
+        exact_depth: The desired length of the path
+        max_depth: The maximum depth of the path
+        use_heuristic: Flag to use potentially erroneous heuristics to limit search space
+
+    Returns:
+        A list of valid word paths or an empty list if we hit a dead end
+
+    Raises:
+        ValueError: If the lengths of the word args are different
+
+    See Also:
+        ladder: Find adjacent words
+        find_ladders_pretty: Print all unique ladders between two words nicely
+    """
+    if len(left_word) != len(right_word):
+        raise ValueError("Word ladder words must be same length")
+
+    cached_searches = {}
+    return follow_ladder_path([left_word.lower()], right_word.lower(), cached_searches, 2, exact_depth, max_depth, use_heuristic)
+
+def find_ladders_pretty(left_word: str, right_word: str, exact_depth: Optional[int] = None, max_depth: int = 10, use_heuristic: bool = True):
+    """ Print unique word ladders (see find_ladders)
+
+    See Also:
+        ladder: Find adjacent words
+        find_ladders: Find all ladders between two words
+    """
+    # Find all ladders and sort by length
+    found_ladders = find_ladders(left_word, right_word, exact_depth, max_depth, use_heuristic)
+    if len(found_ladders) == 0:
+        return
+
+    unique_ladders = set(map(lambda x: frozenset(x), found_ladders))
```

Comment:
Every found ladder is created as sets three times here. Can we reduce this repeated computation?

Reply:
This isn't the part that would hurt perf in any meaningful way, so I'm not going to do anything fancy here. I can remove one of them because it's easy